ext/opentelemetry-ext-dbapi/src/opentelemetry/ext/dbapi/__init__.py

```diff
@@ -0,0 +1,140 @@
+# Copyright 2019, OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+The opentelemetry-ext-dbapi package allows tracing queries made by the
+ibraries following Ptyhon Database API specification:
+https://www.python.org/dev/peps/pep-0249/
+"""
+
+import typing
+
+import wrapt
+
+from opentelemetry.trace import SpanKind, Tracer
+from opentelemetry.trace.status import Status, StatusCanonicalCode
+
+QUERY_WRAP_METHODS = ["execute", "executemany", "callproc"]
+
+
+def trace_integration(
+    tracer: Tracer,
+    connect_module: typing.Callable[..., any],
+    connect_method_name: str,
+    database_component: str,
+):
+    """Integrate with DB API library.
+       https://www.python.org/dev/peps/pep-0249/
+    """
+    # pylint: disable=unused-argument
+    def wrap_connect(
+        wrapped: typing.Callable[..., any],
+        instance: typing.Any,
+        args: typing.Tuple[any, any],
+        kwargs: typing.Dict[any, any],
+    ):
+        db_integration = DatabaseApiIntegration(tracer, database_component)
+        return db_integration.wrapped_connection(wrapped, args, kwargs)
+
+    wrapt.wrap_function_wrapper(
+        connect_module, connect_method_name, wrap_connect
+    )
+
+
+class DatabaseApiIntegration:
+    def __init__(
+        self, tracer: Tracer, database_component: str, database_type="sql"
+    ):
+        if tracer is None:
+            raise ValueError("The tracer is not provided.")
+        self._tracer = tracer
+        self._database_component = database_component
+        self._database_type = database_type
+        self._connection_props = {}
+
+    def wrapped_connection(
+        self,
+        wrapped: typing.Callable[..., any],
+        args: typing.Tuple[any, any],
+        kwargs: typing.Dict[any, any],
+    ):
+        """Patch connection cursor to add tracing.
+        """
+        connection = wrapped(*args, **kwargs)
+        self._connection_props = {
+            "database": connection.database,
+            "port": connection.server_port,
+            "host": connection.server_host,
+            "user": connection.user,
+        }
+
+        wrapt.wrap_function_wrapper(connection, "cursor", self.wrapped_cursor)
```

Comment:
Can `wrap_function_wrapper` ever throw an exception?

Reply:
I guess any call to some external library could throw an exception at some point, are you suggesting to try catch the entire integration to not crash the user app?

Reply:
I'm not sure about catching the entire integration, I am just looking at what [OpenCensus](https://github.com/census-instrumentation/opencensus-python/blob/master/contrib/opencensus-ext-dbapi/opencensus/ext/dbapi/trace.py#L33) does. We don't want to crash the users app like you said which is what OC tries to prevent with the generic try/catch.

Reply:
Added try/except to avoid app to crash if integration fails